### PR TITLE
WIP: #28297 Resolve Same queryset result in two different queries on ORM

### PR DESCRIPTION
**Pull request by orf**
[Ticket](https://code.djangoproject.com/ticket/28297)

### CODE REVIEW

File: django/db/models/sql/query.py [hunk 1]
```diff
@@ -918,6 +918,13 @@ def join(self, join, reuse=None):
         """
         reuse = [a for a, j in self.alias_map.items()
                  if (reuse is None or a in reuse) and j == join]
+
+        if len(reuse) > 1:
+            # In some cases more than one suitable alias can be found. Join.__eq__
+            # only compares on table_name, join_field and parent_alias. Here we find
+            # the alias with a matching join_type.
+            reuse = [r for r in reuse if self.alias_map[r].join_type == join.join_type]
```
Comment (charettes) [inline django/db/models/sql/query.py]: Shouldn't we adjust `Join.__eq__` to consider `join_type` instead? Or maybe move this condition to the reuse generation above?

**Comment by orf** [comment]
I did this in the __eq__, but a *lot* of tests failed with differing
values. I could shift it up into the loop above and see what happens.

On 12 Jun 2017 17:54, "Simon Charette" <notifications@github.com> wrote:

*@charettes* commented on this pull request.
------------------------------

In django/db/models/sql/query.py
<https://github.com/django/django/pull/8631#discussion_r121466856>:

> @@ -918,6 +918,13 @@ def join(self, join, reuse=None):
         """
         reuse = [a for a, j in self.alias_map.items()
                  if (reuse is None or a in reuse) and j == join]
+
+        if len(reuse) > 1:
+            # In some cases more than one suitable alias can be
found. Join.__eq__
+            # only compares on table_name, join_field and
parent_alias. Here we find
+            # the alias with a matching join_type.
+            reuse = [r for r in reuse if self.alias_map[r].join_type
== join.join_type]

Shouldn't we adjust Join.__eq__ to consider join_type instead? Or maybe
move this condition to the reuse generation above?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/django/django/pull/8631#pullrequestreview-43497733>,
or mute
the thread

File: django/db/models/sql/query.py (hunk 1 above)
Comment (orf) [inline django/db/models/sql/query.py]: I did this in the __eq__, but a *lot* of tests failed with differing
values. I could shift it up into the loop above and see what happens.

File: tests/queries/models.py [hunk 2]
```diff
@@ -704,3 +704,20 @@ class RelatedIndividual(models.Model):
 
     class Meta:
         db_table = 'RelatedIndividual'
+
+
+class Ticket28297Recipe(models.Model):
```
Comment (timgraham) [inline tests/queries/models.py]: There's no need to put the ticket number in the model name. If possible to reuse any existing models, that's ideal.

File: tests/queries/tests.py [hunk 3]
```diff
@@ -82,6 +83,19 @@ def setUpTestData(cls):
         Cover.objects.create(title="first", item=i4)
         Cover.objects.create(title="second", item=cls.i2)
 
+    def test_ticket28297(self):
+        # https://code.djangoproject.com/ticket/28297
+        ingredients = ['tomato']
```
Comment (timgraham) [inline tests/queries/tests.py]: I don't see a need for the `ingredients` variable.

File: tests/queries/tests.py [hunk 4]
```diff
@@ -82,6 +83,19 @@ def setUpTestData(cls):
...
+        queryset = Ticket28297Recipe.objects.all()
+        queryset = queryset.annotate(total=Count('steps__ingredients', distinct=True))
+        queryset = queryset.filter(steps__ingredients__ingredient__name__in=ingredients)
+        queryset = queryset.annotate(available=Count('steps__ingredients', distinct=True))
+
```
Comment (timgraham) [inline tests/queries/tests.py]: Chop the blank lines

File: django/db/models/sql/query.py [hunk 5]
```diff
@@ -918,6 +918,13 @@ def join(self, join, reuse=None):
         """
         reuse = [a for a, j in self.alias_map.items()
                  if (reuse is None or a in reuse) and j == join]
+
+        if len(reuse) > 1:
+            # In some cases more than one suitable alias can be found. Join.__eq__
+            # only compares on table_name, join_field and parent_alias. Here we find
```
Comment (timgraham) [inline django/db/models/sql/query.py]: Avoid use of "we" in comments. "Find the alias...." is simpler.

File: django/db/models/sql/query.py (hunk 1 above)
Comment (orf) [inline django/db/models/sql/query.py]: I tried to add it to the `reuse` list comprehension above, but it caused multiple other tests to fail. I did run the whole test suite and no tests seemed to trigger a query that resulted in duplicate aliases, so 🤷‍♂️

**Comment by orf** [comment]
I've fixed up the MR as asked @timgraham. Thank you for the review!

File: tests/queries/tests.py [hunk 6]
```diff
@@ -1172,6 +1172,16 @@ def test_lookup_constraint_fielderror(self):
         with self.assertRaisesMessage(FieldError, msg):
             Tag.objects.filter(unknown_field__name='generic')
 
+    def test_ticket_28297(self):
+        # https://code.djangoproject.com/ticket/28297
```
Comment (timgraham) [inline tests/queries/tests.py]: As discussed in [Coding style](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/#python-style), use a descriptive docstring. The seems complicated enough that a ticket reference is warranted, however, a reader shouldn't have to look up the ticket to understand the basics of the test.
(sorry forgot to press "submit" on this comment during the first review)

Comment (orf) [inline tests/queries/tests.py]: Sure thing, I can add more descriptive comments here for sure. I think @charettes comment in the django ticket is correct though, I'd like to dig into this a bit deeper - there is something else at play here I think.

**Comment by timgraham** [comment]
Should we close this since you deassigned yourself from the ticket?

**Comment by orf** [comment]
Yes, sorry, I forgot to do this.

**Pull request closed without merging**